File: erlang/e003.erl
% Find the largest prime factor of a composite number.

% The prime factors of 13195 are 5, 7, 13 and 29.
% What is the largest prime factor of the number 600851475143 ?

-module(e003).
-export([
    main/1,
    pfactor/1
]).

pfactor(N) ->
    pfactor_(N, []).
pfactor_(N, F) ->
    Next = pfactor_next(N, 2),
    if
        Next == N ->
            [N | F];
        true ->
            pfactor_(trunc(N / Next), [Next | F])
    end.

pfactor_next(N, Factor) ->
    if
        Factor == N ->
            Factor;
        N rem Factor == 0 ->
            Factor;
        true ->
            pfactor_next(N, Factor + 1)
    end.


main(_) ->
    io:format("Prime Factors of 13195: ~w~n", [pfactor(13195)]),
    io:format("Prime Factors of 600851475143: ~w~n", [pfactor(600851475143)]).

File: haskell/e003.hs
{-
Find the largest prime factor of a composite number.

The prime factors of 13195 are 5, 7, 13 and 29.
What is the largest prime factor of the number 600851475143 ?
-}

import Text.Printf

pfactor :: (Integral a) => a -> [a]
pfactor n = pfactor_ n []

pfactor_ :: (Integral a) => a -> [a] -> [a]
pfactor_ n factors = do
    let next = pfactor_next n 2
    if next == n
        then factors ++ [n]
        else pfactor_ (n `div` next) (factors ++ [next])

pfactor_next :: (Integral a) => a -> a -> a
pfactor_next n factor
    | n == factor = factor
    | n `rem` factor == 0 = factor
    | otherwise = pfactor_next n (factor + 1)

main = do
    printf "Largest prime factor of 600851475143: %d\n" (last (pfactor 600851475143) :: Int)
